pgjdbc/src/main/java/org/postgresql/core/ParameterContext.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2020, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.core;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ParameterContext {
+  public static final ParameterContext EMPTY_CONTEXT = new ParameterContext();
```

Comment:
The ParameterContext object here is actually mutable, right? That seems potentially problematic to use as a constant.

Reply:
I have made an attempt to make the object immutable by using the immutable NO_PLACEHOLDERS-list instance.

Reply:
having getPlaceholderPositions return an empty constant, does not make EMPTY_CONTEXT immutable.
The add* methods could still be called (erroneously) on EMPTY_CONTEXT, poisoning constant for the classpath/jvm. Consider an anonymous class which throws exceptions on mutating methods:

```
public static final ParamterContext EMPTY_CONTEXT = new ParamterContext() {
  public int addPositionalParameter(int position) throws SQLException {
    throw new UnsupportedOperationException();
  }

  public int addNamedParameter(int position, String bindName) throws SQLException {
    throw new UnsupportedOperationException();
  }
};
```

Reply:
Ah - now I see what you did in the static initializer. That would seem to make the current implementation immutable (that list is added to first in the current add methods), but would be non-obvious and at risk for being unintentionally refactored.

Reply:
I guess that is why we have tests, right? But anyway I think your solution is better.